### PR TITLE
Feat(web-twig): Introduce Pagination component #DS-328

### PR DESCRIPTION
**Pull request by crishpeen**

### CODE REVIEW

File: packages/web-twig/src/Resources/components/Pagination/PaginationLinkPrevious.twig [hunk 1]
```diff
@@ -0,0 +1,22 @@
+{# API #}
+{%- set props = props | default([]) -%}
+{%- set _href = props.href -%}
+{%- set _label = props.label | default('Previous') -%}
+{%- set _target = props.target | default(null) -%}
+
+{# Miscellaneous #}
+{%- set _styleProps = useStyleProps(props) -%}
+
+{% embed "@spirit/buttonLink.twig" with { props: {
```
Comment (literat) [inline packages/web-twig/src/Resources/components/Pagination/PaginationLinkPrevious.twig:10]: You cannot use TwigX syntax here?

**Review by literat — CHANGES_REQUESTED** [review]

File: packages/web-twig/src/Resources/components/Pagination/PaginationLinkPrevious.twig (hunk 1 above)
Comment (crishpeen) [reply packages/web-twig/src/Resources/components/Pagination/PaginationLinkPrevious.twig:10]: nope, I am not able to get there mainProps.

File: packages/web-twig/src/Resources/components/Pagination/PaginationWrapper.twig [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+{%- set props = props | default([]) -%}
+
+{# Miscellaneous #}
+{%- set _styleProps = useStyleProps(props) -%}
+{%- set _classNames = [ _styleProps.className ] -%}
+
+<nav
+    {{ mainProps(props) }}
+    {{ styleProp(_styleProps) }}
+    {{ classProp(_classNames) }}
```
Comment (literat) [inline packages/web-twig/src/Resources/components/Pagination/PaginationWrapper.twig:11]: We can use this directly, can't we?

Comment (crishpeen) [reply packages/web-twig/src/Resources/components/Pagination/PaginationWrapper.twig:11]: Generally, we can, but this adds two features out of the box. 
1. it makes adding new classes easy
2. it checks if there are any values and outputs the class attribute only if so

And it was easier for me to implement 😁 
If you insist on direct usage, I can di refactoring.

File: packages/web-twig/src/Resources/components/Pagination/Pagination.stories.twig [hunk 3]
```diff
@@ -0,0 +1,17 @@
+{% extends 'layout/plain.html.twig' %}
+
+{% block content %}
+
+{# Pagination Current First #}
+{% include '@components/Pagination/stories/PaginationCurrentFirst.twig' %}
+
+{# Pagination Current First #}
```
Comment (tomassychra) [inline packages/web-twig/src/Resources/components/Pagination/Pagination.stories.twig:8]:
```suggestion
{# Pagination Current Middle #}
```

File: packages/web-twig/src/Resources/components/Pagination/Pagination.stories.twig [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+{% include '@components/Pagination/stories/PaginationCurrentMiddle.twig' %}
+
+{# Pagination Current First #}
```
Comment (tomassychra) [inline packages/web-twig/src/Resources/components/Pagination/Pagination.stories.twig:11]:
```suggestion
{# Pagination Current Last #}
```

File: packages/web-twig/src/Resources/components/Pagination/Pagination.stories.twig [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+{% include '@components/Pagination/stories/PaginationCurrentLast.twig' %}
+
+{# Pagination Current First #}
```
Comment (tomassychra) [inline packages/web-twig/src/Resources/components/Pagination/Pagination.stories.twig:14]:
```suggestion
{# Pagination Current First Centered #}
```

**Comment by tomassychra** [comment]
I was wondering how to access an element with the `accessibility-hidden` class. Since it is repeated everywhere with the same value, I prepared it for the React component as a label property with the possibility of overwriting. I am considering an even better name `hiddenLabel`. What do you think about it? @crishpeen @literat @adamkudrna

File: packages/web-twig/src/Resources/components/Pagination/PaginationLinkNext.twig [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    color: 'secondary',
+    href: _href,
+    isSquare: true,
+    target: _target,
+    UNSAFE_className: _styleProps.className,
+    UNSAFE_style: _styleProps.style,
+}) } %}
+    {% block content %}
+        <Icon name="chevron-right" />
+        <span class="accessibility-hidden">{{ _label }}</span>
```
Comment (tomassychra) [inline packages/web-twig/src/Resources/components/Pagination/PaginationLinkNext.twig:21]:
```suggestion
        <span class="accessibility-hidden">{{ _hiddenLabel }}</span>
```
What do you think?

File: packages/web-twig/src/Resources/components/Pagination/PaginationLinkPrevious.twig [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    color: 'secondary',
+    href: _href,
+    isSquare: true,
+    target: _target,
+    UNSAFE_className: _styleProps.className,
+    UNSAFE_style: _styleProps.style,
+}) } %}
+    {% block content %}
+        <Icon name="chevron-left" />
+        <span class="accessibility-hidden">{{ _label }}</span>
```
[the remark this answers is not in the document]
Comment (literat) [reply packages/web-twig/src/Resources/components/Pagination/PaginationLinkPrevious.twig:21]: @tomassychra @crishpeen I have checked our code for usage of `accessibility-hidden` and `label` and there is no usage of the`hiddenLabel`. We use a `closeLabel` for the close button cross or just a `label`. In this case, I would prefer to leave just `label` or use `previousLabel` or `nextLabel`.

**Review by literat — APPROVED** [review]

File: packages/web-twig/src/Resources/components/Pagination/PaginationLinkPrevious.twig (hunk 7 above)
[the remark this answers is not in the document]
Comment (tomassychra) [reply packages/web-twig/src/Resources/components/Pagination/PaginationLinkPrevious.twig:21]: But the problem here is that the normal label is visible, this "label" is not visible, I have already renamed it. I think it's just a confusing name.

**Review by tomassychra — APPROVED** [review]

File: packages/web-twig/src/Resources/components/Pagination/stories/PaginationCurrentMiddle.twig [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    </PaginationItem>
+    <PaginationItem>
+        <PaginationLink
+            onclick="console.log('page 15')"
+            accessibilityLabel="Goto Page 15"
+            pageNumber="15"
+        />
+    </PaginationItem>
+    <PaginationItem>
+        <PaginationLinkNext onclick="console.log('next')" />
```
[the remark this answers is not in the document]
Comment (crishpeen) [reply packages/web-twig/src/Resources/components/Pagination/stories/PaginationCurrentMiddle.twig:42]: you are right, I will fix it in the docs

File: packages/web-twig/tests/__snapshots__/ComponentsSnapshotTest__test with data set paginationDefault.twig__1.html [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+  </head>
+  <body>
+    <nav role="navigation">
+      <ul class="Pagination">
+        <li class="Pagination__item">
+          <a class="Pagination__link Pagination__link--current" href="&amp;#x23;" aria-current="page"><span class="accessibility-hidden">Current Page, Page 1</span> <span aria-hidden="true">1</span></a>
+        </li>
+
+        <li class="Pagination__item">
+          <a class="Pagination__link" href="&amp;#x23;"><span class="accessibility-hidden">Goto Page 2</span>
```
Comment (adamkudrna) [inline packages/web-twig/tests/__snapshots__/ComponentsSnapshotTest__test with data set paginationDefault.twig__1.html:15]: Is it really "Goto", not "Go to"…?

File: packages/web-twig/src/Resources/components/Pagination/README.md [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+  </PaginationItem>
+</Pagination>
+```
+
+#### API
+
+| Prop name            | Type     | Default              | Required | Description                       |
+| -------------------- | -------- | -------------------- | -------- | --------------------------------- |
+| `accessibilityLabel` | `string` | `Next` or `Previous` | no       | Accessibility label of the button |
+| `href`               | `string` | —                    | yes      | Link URL                          |
```
Comment (adamkudrna) [inline packages/web-twig/src/Resources/components/Pagination/README.md:144]:
```suggestion
| `href`               | `string` | —                    | no      | Link URL                          |
```

Actually, `href` is optional.

File: packages/web-twig/src/Resources/components/Pagination/README.md [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+descriptiveness and accessibility. Also, UNSAFE styling props are available,
+see the [Escape hatches][escape-hatches] section in README to learn how and when to use them.
+These attributes will be passed to the topmost HTML element of the component.
+
+### Full Example
+
+When you put it all together:
+
+```twig
+<Pagination label="Page navigation">
```
Comment (adamkudrna) [inline packages/web-twig/src/Resources/components/Pagination/README.md:157]:
```suggestion
<Pagination aria-label="Page navigation">
```

File: packages/web-twig/tests/components-fixtures/paginationDefault.twig [hunk 12]
```diff
@@ -0,0 +1,41 @@
+<Pagination label="Page navigation">
```
Comment (adamkudrna) [inline packages/web-twig/tests/components-fixtures/paginationDefault.twig:1]:
```suggestion
<Pagination aria-label="Page navigation">
```

**Review by adamkudrna — APPROVED** [review]
Nice! 👍🏻

File: packages/web-twig/src/Resources/components/Pagination/Pagination.stories.twig [hunk 13]
```diff
@@ -0,0 +1,17 @@
+{% extends 'layout/plain.html.twig' %}
+
+{% block content %}
+
+{# Pagination Current First #}
+{% include '@components/Pagination/stories/PaginationCurrentFirst.twig' %}
```
Comment (adamkudrna) [inline packages/web-twig/src/Resources/components/Pagination/Pagination.stories.twig:6]: Maybe the stories could have headlines like in [the React implementation](https://deploy-preview-879--spirit-design-system-react.netlify.app/src/components/pagination/)?

obrazek

File: packages/web-twig/tests/__snapshots__/ComponentsSnapshotTest__test with data set paginationDefault.twig__1.html (hunk 9 above)
Comment (literat) [reply packages/web-twig/tests/__snapshots__/ComponentsSnapshotTest__test with data set paginationDefault.twig__1.html]: According to many dictionaries, it is a "Go to". I fixed that.

Comment (literat) [reply packages/web-twig/tests/__snapshots__/ComponentsSnapshotTest__test with data set paginationDefault.twig__1.html]: Please see also https://github.com/lmc-eu/spirit-design-system/pull/895 where I fixed this for the web package.

File: packages/web-twig/src/Resources/components/Pagination/README.md (hunk 10 above)
Comment (literat) [reply packages/web-twig/src/Resources/components/Pagination/README.md]: Ok, fixed.

File: packages/web-twig/src/Resources/components/Pagination/Pagination.twig [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+{%- set _listClassName = _spiritClassPrefix ~ 'Pagination' -%}
+
+{# Miscellaneous #}
+{%- set _styleProps = useStyleProps(props) -%}
+
+<nav
+    {{ mainProps(props) }}
+    {{ styleProp(_styleProps) }}
+    {{ classProp( [ _styleProps.className ] ) }}
+    role="navigation"
```
[the remark this answers is not in the document]
Comment (literat) [reply packages/web-twig/src/Resources/components/Pagination/Pagination.twig]: Ok, fixed.

File: packages/web-twig/tests/components-fixtures/paginationDefault.twig (hunk 12 above)
Comment (literat) [reply packages/web-twig/tests/components-fixtures/paginationDefault.twig:1]: Fixed.

File: packages/web-twig/src/Resources/components/Pagination/stories/PaginationCurrentFirst.twig [hunk 15]
```diff
@@ -0,0 +1,41 @@
+<Pagination UNSAFE_className="mb-800" label="Page navigation">
```
[the remark this answers is not in the document]
Comment (literat) [reply packages/web-twig/src/Resources/components/Pagination/stories/PaginationCurrentFirst.twig]: Fixed.

File: packages/web-twig/src/Resources/components/Pagination/README.md (hunk 11 above)
Comment (literat) [reply packages/web-twig/src/Resources/components/Pagination/README.md]: Fixed.

File: packages/web-twig/src/Resources/components/Pagination/Pagination.stories.twig (hunk 13 above)
Comment (literat) [reply packages/web-twig/src/Resources/components/Pagination/Pagination.stories.twig]: Fixed that.

File: packages/web-twig/tests/__snapshots__/ComponentsSnapshotTest__test with data set paginationDefault.twig__1.html [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+          <span aria-hidden="true">2</span></a>
+        </li>
+
+        <li class="Pagination__item">
+          <a class="Pagination__link" href="&amp;#x23;"><span class="accessibility-hidden">Goto Page 3</span>
```
[the remark this answers is not in the document]
Comment (literat) [reply packages/web-twig/tests/__snapshots__/ComponentsSnapshotTest__test with data set paginationDefault.twig__1.html]: Fixed that.

File: packages/web-twig/src/Resources/components/Pagination/stories/PaginationCurrentMiddle.twig (hunk 8 above)
[the remark this answers is not in the document]
Comment (literat) [reply packages/web-twig/src/Resources/components/Pagination/stories/PaginationCurrentMiddle.twig:42]: Fixed.

File: packages/web-twig/src/Resources/components/Pagination/Pagination.stories.twig [hunk 17]
```diff
@@ -0,0 +1,29 @@
+{% extends 'layout/plain.html.twig' %}
+
+{% block content %}
+
+{# Pagination Current First #}
+<section class="docs-Section">
+    <h2 class="docs-Heading">Pagination Current First</h2>
+    {% include '@components/Pagination/stories/PaginationCurrentFirst.twig' %}
+</section>
```
Comment (adamkudrna) [inline packages/web-twig/src/Resources/components/Pagination/Pagination.stories.twig:9]: Thanks for the headings! Just one more thing: we always have the section with heading inside the included story… Would you fix that please? 🙏🏻

**Review by adamkudrna — APPROVED** [review]

File: packages/web-twig/src/Resources/components/Pagination/Pagination.stories.twig (hunk 17 above)
Comment (literat) [reply packages/web-twig/src/Resources/components/Pagination/Pagination.stories.twig:9]: That is not true. We want to use the stories in the future for rendering in Supernova or for snapshots tests. In that case, the heading will be a problem, I think.

**Pull request merged**